src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,96 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] values;
+    private int size;
+
+    public ArrayList() {
+        values = new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (values.length == size) {
+            T[] sourceArray = (T[]) values;
+            values = new Object[size + (size >> 1)];
+            System.arraycopy(sourceArray, 0, values, 0, size);
+        }
+        values[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index:" + index + "out of size" + size);
+        }
+        if (values.length == size) {
+            T[] sourceArray = (T[]) values;
+            values = new Object[size + (size >> 1)];
+            System.arraycopy(sourceArray, 0, values, 0, size);
+        }
+        System.arraycopy(values, index, values, index + 1, size - index);
+        values[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index:" + index + "out of size" + size);
+        }
```

Comment:
Common [mistakes](https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/array-list.html#dont-create-repeating-code)
**Don’t create repeating code.**
If the logic of your code repeats - move it to the separate private method. For example, the index check for get(), set() and remove() methods should be moved to the separate method.